### PR TITLE
fix: Need to wait for zip deployment to be done processing

### LINKED ISSUE #7809: Exporting node WA as skills throws deploy error
## Describe the bug
Exporting node web apps as skills has a zip deploy error for invalid hostname

## Version
Version:  2.0.0-nightly.245007.8406071
Electron: 8.2.4
Chrome: 80.0.3987.165
NodeJS: 12.13.0
V8: 8.0.426.27-electron.0

## Browser

- [x] Electron distribution
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [ ] macOS
- [x] Windows
- [ ] Ubuntu

## To Reproduce

1. Create an node empty bot web app bot
2. Create a provisioning profile for the bot
3. While the provisioning is going on add an echo trigger and dialog
4. After the provision finishes publish the bot
5. once the bot is published export the bot as a skill
6. only select the echo trigger and dialog

## Expected behavior
This should work

## Screenshots
image

image

## Additional context
There's other issues with deploying node bots, but this is the only one where the zip deploy error is thrown.

I should add I was able to repo this with a fresh bot to rule out Composer being in some weird state. Here's the log from Composer:

---
"Accepted for publishing..."
---

---
"Abs settings incomplete, skip linking bot with webapp ..."
---

---
"appPasswordHint incomplete, return ..."
---

---
"Building the bot app..."
---

---
"Build succeeded!"
---

---
"start publish luis"
---

---
"Assigning to luis app id: c3937102-19a2-42e9-b7ba-c9f9e6ca1539"
---

---
"Luis Publish Success! ..."
---

---
"Creating build artifact..."
---

---
"Build artifact ready!"
---

---
"Publishing to Azure ..."
---

---
"Uploading zip file... to NWAEcho513z"
---

---
"{\"name\":\"DEPLOY_ZIP_ERROR\",\"message\":\"The hostname is invalid, please check if hostname/name-environment matches your target resource(webapp/function) name\",\"stack\":\"Error\\n    at new CustomizeError (C:\\\\Users\\\\joratlif\\\\AppData\\\\Local\\\\Programs\\\\Bot Framework Composer\\\\resources\\\\app.asar.unpacked\\\\extensions\\\\azurePublish\\\\dist\\\\extension.js:255547:25)\\n    at createCustomizeError2 (C:\\\\Users\\\\joratlif\\\\AppData\\\\Local\\\\Programs\\\\Bot Framework Composer\\\\resources\\\\app.asar.unpacked\\\\extensions\\\\azurePublish\\\\dist\\\\extension.js:255552:10)\\n    at BotProjectDeploy.deployZip (C:\\\\Users\\\\joratlif\\\\AppData\\\\Local\\\\Programs\\\\Bot Framework Composer\\\\resources\\\\app.asar.unpacked\\\\extensions\\\\azurePublish\\\\dist\\\\extension.js:256037:15)\"}"
---

---
"DEPLOY_ZIP_ERROR - The hostname is invalid, please check if hostname/name-environment matches your target resource(webapp/function) name"
---

### PR DESCRIPTION
**Pull request by GeoffCoxMSFT**
## Description

As discovered as part of R13, the zip deploy call assumes deployment is done when the method returns.  
Instead the publishing process needs to poll for status when the method returns a 202 (in progress).
This updates the deploy zip call to poll every 5 seconds when the initial call result is 202.

## Task Item

fixes #7809 

## Screenshots

Here is a log output:

---
"Publishing to Azure ..."
---

---
"Uploading zip file... to gcoxResource2"
---

---
"Waiting for zip upload processing."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Receiving changes."
---

---
"Waiting for zip upload processing. Building and Deploying 'ff3552b2581b48208e595dbd18cfdf88'."
---

---
"Waiting for zip upload processing. Building and Deploying 'ff3552b2581b48208e595dbd18cfdf88'."
---

---
"Zip upload processed successfully."
---

---
"Published successfully!"

### CODE REVIEW

**Comment by GeoffCoxMSFT** [comment]
@cwhitten I think this should go into the patch release of Composer coming up.

File: extensions/azurePublish/src/node/deploy.ts [hunk 1]
```diff
@@ -238,6 +238,42 @@ export class BotProjectDeploy {
...
+  private async waitForZipDeploy(token: string, statusUrl: string): Promise<void> {
+    this.logger({
+      status: BotProjectDeployLoggerType.DEPLOY_INFO,
+      message: `Waiting for zip upload processing.`,
+    });
+
+    return new Promise((resolve, reject) => {
+      const timerId = setInterval(async () => {
+        try {
+          const statusResponse = await axios.get(statusUrl, { headers: { Authorization: `Bearer ${token}` } });
```
Comment (cwhitten) [inline extensions/azurePublish/src/node/deploy.ts:250]: I have an ask out to @tonyanziano on of this token can expire while polling here

**Review by cwhitten — APPROVED** [review]
One edge-case I think we should clarify, otherwise LGTM

File: extensions/azurePublish/src/node/deploy.ts (hunk 1 above)
Comment (tonyanziano) [reply extensions/azurePublish/src/node/deploy.ts:250]: After looking at the code it looks like ARM tokens _can_ expire during long tasks, and we **do not** refresh these automatically behind the scenes.

This is something we do [for regular access tokens](https://github.com/microsoft/BotFramework-Composer/blob/main/Composer/packages/server/src/services/auth/electronAuthProvider.ts#L124), but it has not been implemented in the [ARM token flow](https://github.com/microsoft/BotFramework-Composer/blob/main/Composer/packages/server/src/services/auth/electronAuthProvider.ts#L63).

I think the reasoning behind this was that logging in to fetch an ARM token usually requires an interactive sign-in and so the UX would not be great for the user if they were suddenly shown a login window without any context when refreshing a token. That being said, I don't think we ever tested this hypothesis out.

We could try using the same token refresh logic with ARM tokens and see what happens, but again, I'm unclear on how the UX would look.

Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/node/deploy.ts:250]: During R14 we're revamping the service layer for prov/pub a bit and abstracting the polling for long running operations.  I'll ask Natalie to consider the options for refreshing the token while waiting.

Comment (cwhitten) [reply extensions/azurePublish/src/node/deploy.ts:250]: Awesome - sounds great

**Pull request merged**